File: Dart/lib/Experimenter.dart
import 'package:Dart/ExperimentResults.dart';
import 'package:Dart/Problem.dart';

class Experimenter {

  ExperimentResults RunExperiment(Problem p){
    print("Running Experiment");
    List<double> measurements = [].toList();
    bool res = false;
    Stopwatch sw = new Stopwatch();

    for (int i = 0; i<100000; i++){

      sw.reset();
      sw.start();
      res = p.Calculate();
      sw.stop();
      int ticks = sw.elapsedTicks;
      measurements.add(ticks/sw.frequency * 1000000000);
      if (!res){
        print("Algorithm failed\n");
      }

    }

    return new ExperimentResults(measurements, Average(measurements), Median(measurements), High(measurements), Low(measurements));
  }


  double Average(List<double> list){
    double total = 0.0;
    for (double x in list){
      total+= x;
    }
    return total/list.length;
  }

  double Median(List<double> list){
    list.sort();
    return list[list.length~/2];
  }

  double High(List<double> list){
    list.sort();
    return list[list.length-1];
  }

  double Low(List<double> list){
    list.sort();
    return list[0];
  }

  void PrintIndi(List<ExperimentResults> results){
    for (ExperimentResults x in results){
      x.PrintMe();
    }
  }

  void PrintAverages(List<ExperimentResults> results){
    print("Printing Averages: ");
    print("Average-Average: "+AvgOfAvg(results).round().toString()+"ns");
    print("Average-Median: "+AvgOfMedian(results).round().toString()+"ns");
    print("Average-High: "+AvgOfHigh(results).round().toString()+"ns");
    print("Average-Low: "+AvgOfLow(results).round().toString()+"ns");
  }

  double AvgOfAvg(List<ExperimentResults> results){
    double total = 0.0;
    for (ExperimentResults x in results){
      total += x.avg;
    }
    return total/results.length;
  }

  double AvgOfMedian(List<ExperimentResults> results){
    double total = 0.0;
    for (ExperimentResults x in results){
      total += x.med;
    }
    return total/results.length;
  }

  double AvgOfHigh(List<ExperimentResults> results){
    double total = 0.0;
    for (ExperimentResults x in results){
      total += x.high;
    }
    return total/results.length;
  }

  double AvgOfLow(List<ExperimentResults> results){
    double total = 0.0;
    for (ExperimentResults x in results){
      total += x.low;
    }
    return total/results.length;
  }





}

File: PHP/Experimenter.php
<?php
/**
 * Created by PhpStorm.
 * User: Animc
 * Date: 28-07-2018
 * Time: 12:57
 */

require 'Problem.php';

class Experimenter
{

    public function RunExperiment(Problem $problem){

        $mes = [];
        for ($i = 0; $i < 1000000; $i++){
            $start = microtime(true);
            $res = $problem->Calculate();
            $time_elapsed = microtime(true) - $start;
            $nanos = $time_elapsed*1000000000;
            array_push($mes, $nanos);
            if (!$res){
                print "Algorithm failed==================================\nAlgorithm failed==================================\nAlgorithm failed==================================\nAlgorithm failed==================================\nAlgorithm failed==================================\n";
                break;
            }
        }
        return new ExperimentResults(get_class($problem), $mes, $this->Average($mes),$this->Median($mes),$this->High($mes),$this->Low($mes));
    }


    function Average($numbers){
        $total = 0;
        foreach ($numbers as $value){
            $total = $total + $value;
        }
        return $total/count($numbers);
    }

    function Median($numbers){
        sort($numbers);
        return $numbers[count($numbers)/2];

    }

    function High($numbers){
        sort($numbers);
        return $numbers[count($numbers)-1];
    }

    function Low($numbers){
        sort($numbers);
        return $numbers[0];
    }

    public function PrintIndi(array $results){
        foreach ($results as $result){
            ExperimentResults::PrintMe($result);
        }
    }

    public function PrintAverages(array $results){
        print "Printing Averages on: ".ExperimentResults::GetName($results[0])."\n";
        print "Average-Average: ".round($this->AvgOfAvg($results))."ns\n";
        print "Average-Median: ".round($this->AvgOfMed($results))."ns\n";
        print "Average-High: ".round($this->AvgOfHigh($results))."ns\n";
        print "Average-Low: ".round($this->AvgOfLow($results))."ns\n";
    }

    function AvgOfAvg(array $results){
        $total = 0;
        foreach ($results as $result){
            $total = $total+ExperimentResults::GetAvg($result);
        }
        return $total/count($results);
    }

    function AvgOfMed(array $results){
        $total = 0;
        foreach ($results as $result){
            $total = $total+ExperimentResults::GetMed($result);
        }
        return $total/count($results);
    }

    function AvgOfHigh(array $results){
        $total = 0;
        foreach ($results as $result){
            $total = $total+ExperimentResults::GetHigh($result);
        }
        return $total/count($results);
    }

    function AvgOfLow(array $results){
        $total = 0;
        foreach ($results as $result){
            $total = $total+ExperimentResults::GetLow($result);
        }
        return $total/count($results);
    }

}